### DISCUSSION TITLE
define routing to a class method?

### DISCUSSION THREAD

**Discussion by erhuabushuo** [emmett-framework/emmett#382] [Q&A] [answered]
define routing to a class method?

**Comment by gi0baro** [comment]
@erhuabushuo by "class method" you mean a method defined into a class decorated by `classmethod` or an instance method?

The first is not supported, but the second one.
I'm not sure why would you do that, though. Can you explain your use-case?

**Comment by josejachuf** [comment]
So I understand it would be like that:

```
@app.route('/mymethod', methods='get')
async def mymethod():
    return MyClass.my_method()
```

Jose

**Comment by erhuabushuo** [comment]
like flask class based view

**Comment by gi0baro** [comment, marked as answer]
@erhuabushuo Emmett doesn't provide any ready-to-go facilities for that, mainly because in Emmett a class based view is an anti-pattern, secondly to prevent wasting performance in creating objects during request flow.

You might want to implement your own `AppModule` class for such logic, somewhat like [the REST extension](https://github.com/emmett-framework/rest) does:

```python
from emmett import AppModule

class ViewModule(AppModule):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._define_routes()

    def _define_routes(self):
        for method in ["get", "post", "put", "delete"]:
            if handler := getattr(self, method, None):
                self.route(methods=[method], name=method)(handler)


class TodoView(ViewModule):
    def get(self):
        return {"key": "value"}


todo = app.module(__name__, "todo", url_prefix="/todo", module_class=TodoView)
```

Mind that the instance is shared across all requests, so avoid to set attributes to `self` inside your methods. You should use `current`, `request` or `session` objects for that.
